// @nrwl/angular/src/generators/init/angular-v14/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.angularInitGenerator = void 0;
const tslib_1 = require("tslib");
const cypress_1 = require("@nrwl/cypress");
const devkit_1 = require("@nrwl/devkit");
const jest_1 = require("@nrwl/jest");
const linter_1 = require("@nrwl/linter");
const run_tasks_in_serial_1 = require("@nrwl/workspace/src/utilities/run-tasks-in-serial");
const backward_compatible_versions_1 = require("../../../utils/backward-compatible-versions");
const test_runners_1 = require("../../../utils/test-runners");
const karma_1 = require("../../karma/karma");
const version_utils_1 = require("../../utils/version-utils");
function angularInitGenerator(host, rawOptions) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const options = normalizeOptions(rawOptions);
        setDefaults(host, options);
        const peerDepsToInstall = [
            '@angular-devkit/core',
            '@angular-devkit/schematics',
            '@schematics/angular',
        ];
        let devkitVersion;
        peerDepsToInstall.forEach((pkg) => {
            var _a;
            const packageVersion = (0, version_utils_1.getInstalledPackageVersion)(host, pkg);
            if (!packageVersion) {
                devkitVersion !== null && devkitVersion !== void 0 ? devkitVersion : (devkitVersion = (_a = (0, version_utils_1.getInstalledPackageVersion)(host, '@angular-devkit/build-angular')) !== null && _a !== void 0 ? _a : backward_compatible_versions_1.backwardCompatibleVersions.angularV14.angularDevkitVersion);
                (0, devkit_1.ensurePackage)(host, pkg, devkitVersion);
            }
        });
        const depsTask = !options.skipPackageJson
            ? updateDependencies(host)
            : () => { };
        const unitTestTask = yield addUnitTestRunner(host, options);
        const e2eTask = addE2ETestRunner(host, options);
        addGitIgnoreEntry(host, '.angular');
        if (!options.skipFormat) {
            yield (0, devkit_1.formatFiles)(host);
        }
        return (0, run_tasks_in_serial_1.runTasksInSerial)(depsTask, unitTestTask, e2eTask);
    });
}
exports.angularInitGenerator = angularInitGenerator;
function normalizeOptions(options) {
    var _a, _b, _c, _d, _e, _f, _g;
    return {
        e2eTestRunner: (_a = options.e2eTestRunner) !== null && _a !== void 0 ? _a : test_runners_1.E2eTestRunner.Cypress,
        linter: (_b = options.linter) !== null && _b !== void 0 ? _b : linter_1.Linter.EsLint,
        skipFormat: (_c = options.skipFormat) !== null && _c !== void 0 ? _c : false,
        skipInstall: (_d = options.skipInstall) !== null && _d !== void 0 ? _d : false,
        skipPackageJson: (_e = options.skipPackageJson) !== null && _e !== void 0 ? _e : false,
        style: (_f = options.style) !== null && _f !== void 0 ? _f : 'css',
        unitTestRunner: (_g = options.unitTestRunner) !== null && _g !== void 0 ? _g : test_runners_1.UnitTestRunner.Jest,
    };
}
function setDefaults(host, options) {
    const nxJson = (0, devkit_1.readNxJson)(host);
    nxJson.generators = nxJson.generators || {};
    nxJson.generators['@nrwl/angular:application'] = Object.assign({ style: options.style, linter: options.linter, unitTestRunner: options.unitTestRunner, e2eTestRunner: options.e2eTestRunner }, (nxJson.generators['@nrwl/angular:application'] || {}));
    nxJson.generators['@nrwl/angular:library'] = Object.assign({ linter: options.linter, unitTestRunner: options.unitTestRunner }, (nxJson.generators['@nrwl/angular:library'] || {}));
    nxJson.generators['@nrwl/angular:component'] = Object.assign({ style: options.style }, (nxJson.generators['@nrwl/angular:component'] || {}));
    (0, devkit_1.updateNxJson)(host, nxJson);
}
function updateDependencies(tree) {
    var _a, _b;
    const angularVersion = (_a = (0, version_utils_1.getInstalledPackageVersion)(tree, '@angular/core')) !== null && _a !== void 0 ? _a : backward_compatible_versions_1.backwardCompatibleVersions.angularV14.angularVersion;
    const angularDevkitVersion = (_b = (0, version_utils_1.getInstalledPackageVersion)(tree, '@angular-devkit/build-angular')) !== null && _b !== void 0 ? _b : backward_compatible_versions_1.backwardCompatibleVersions.angularV14.angularDevkitVersion;
    return (0, version_utils_1.addDependenciesToPackageJsonIfDontExist)(tree, {
        '@angular/animations': angularVersion,
        '@angular/common': angularVersion,
        '@angular/compiler': angularVersion,
        '@angular/core': angularVersion,
        '@angular/forms': angularVersion,
        '@angular/platform-browser': angularVersion,
        '@angular/platform-browser-dynamic': angularVersion,
        '@angular/router': angularVersion,
        rxjs: backward_compatible_versions_1.backwardCompatibleVersions.angularV14.rxjsVersion,
        tslib: backward_compatible_versions_1.backwardCompatibleVersions.angularV14.tsLibVersion,
        'zone.js': backward_compatible_versions_1.backwardCompatibleVersions.angularV14.zoneJsVersion,
    }, {
        '@angular/cli': angularDevkitVersion,
        '@angular/compiler-cli': angularVersion,
        '@angular/language-service': angularVersion,
        '@angular-devkit/build-angular': angularDevkitVersion,
    });
}
function addUnitTestRunner(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        switch (options.unitTestRunner) {
            case test_runners_1.UnitTestRunner.Karma:
                return yield (0, karma_1.karmaGenerator)(tree, {
                    skipPackageJson: options.skipPackageJson,
                });
            case test_runners_1.UnitTestRunner.Jest:
                if (!options.skipPackageJson) {
                    (0, version_utils_1.addDependenciesToPackageJsonIfDontExist)(tree, {}, {
                        'jest-preset-angular': backward_compatible_versions_1.backwardCompatibleVersions.angularV14.jestPresetAngularVersion,
                    });
                }
                return (0, jest_1.jestInitGenerator)(tree, {
                    skipPackageJson: options.skipPackageJson,
                });
            default:
                return () => { };
        }
    });
}
function addE2ETestRunner(tree, options) {
    switch (options.e2eTestRunner) {
        case test_runners_1.E2eTestRunner.Protractor:
            return !options.skipPackageJson
                ? (0, version_utils_1.addDependenciesToPackageJsonIfDontExist)(tree, {}, {
                    protractor: backward_compatible_versions_1.backwardCompatibleVersions.angularV14.protractorVersion,
                    'jasmine-core': backward_compatible_versions_1.backwardCompatibleVersions.angularV14.jasmineCoreVersion,
                    'jasmine-spec-reporter': backward_compatible_versions_1.backwardCompatibleVersions.angularV14
                        .jasmineSpecReporterVersion,
                    'ts-node': backward_compatible_versions_1.backwardCompatibleVersions.angularV14.tsNodeVersion,
                    '@types/jasmine': backward_compatible_versions_1.backwardCompatibleVersions.angularV14.typesJasmineVersion,
                    '@types/jasminewd2': backward_compatible_versions_1.backwardCompatibleVersions.angularV14.typesJasminewd2Version,
                })
                : () => { };
        case test_runners_1.E2eTestRunner.Cypress:
            return (0, cypress_1.cypressInitGenerator)(tree, {
                skipPackageJson: options.skipPackageJson,
            });
        default:
            return () => { };
    }
}
function addGitIgnoreEntry(host, entry) {
    if (host.exists('.gitignore')) {
        let content = host.read('.gitignore', 'utf-8');
        if (/^\.angular$/gm.test(content)) {
            return;
        }
        content = `${content}\n${entry}\n`;
        host.write('.gitignore', content);
    }
    else {
        devkit_1.logger.warn(`Couldn't find .gitignore file to update`);
    }
}
exports.default = angularInitGenerator;
