// @nrwl/angular/src/generators/setup-ssr/lib/update-project-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateProjectConfig = void 0;
const devkit_1 = require("@nrwl/devkit");
function updateProjectConfig(tree, schema) {
    var _a, _b, _c;
    let projectConfig = (0, devkit_1.readProjectConfiguration)(tree, schema.project);
    projectConfig.targets.build.options.outputPath = `dist/apps/${schema.project}/browser`;
    const buildTargetFileReplacements = (_b = (_a = projectConfig.targets.build.configurations) === null || _a === void 0 ? void 0 : _a.production) === null || _b === void 0 ? void 0 : _b.fileReplacements;
    projectConfig.targets.server = {
        dependsOn: ['build'],
        executor: '@angular-devkit/build-angular:server',
        options: {
            outputPath: `dist/${projectConfig.root}/server`,
            main: (0, devkit_1.joinPathFragments)(projectConfig.root, schema.serverFileName),
            tsConfig: (0, devkit_1.joinPathFragments)(projectConfig.root, 'tsconfig.server.json'),
        },
        configurations: {
            production: Object.assign({ outputHashing: 'media' }, (buildTargetFileReplacements
                ? { fileReplacements: buildTargetFileReplacements }
                : {})),
            development: {
                optimization: false,
                sourceMap: true,
                extractLicenses: false,
            },
        },
        defaultConfiguration: 'production',
    };
    projectConfig.targets['serve-ssr'] = {
        executor: '@nguniversal/builders:ssr-dev-server',
        configurations: {
            development: {
                browserTarget: `${schema.project}:build:development`,
                serverTarget: `${schema.project}:server:development`,
            },
            production: {
                browserTarget: `${schema.project}:build:production`,
                serverTarget: `${schema.project}:server:production`,
            },
        },
        defaultConfiguration: 'development',
    };
    projectConfig.targets.prerender = {
        executor: '@nguniversal/builders:prerender',
        options: {
            routes: ['/'],
        },
        configurations: {
            development: {
                browserTarget: `${schema.project}:build:development`,
                serverTarget: `${schema.project}:server:development`,
            },
            production: {
                browserTarget: `${schema.project}:build:production`,
                serverTarget: `${schema.project}:server:production`,
            },
        },
        defaultConfiguration: 'production',
    };
    (0, devkit_1.updateProjectConfiguration)(tree, schema.project, projectConfig);
    const nxJson = (0, devkit_1.readNxJson)(tree);
    if (((_c = nxJson.tasksRunnerOptions) === null || _c === void 0 ? void 0 : _c.default) &&
        !nxJson.tasksRunnerOptions.default.options.cacheableOperations.includes('server')) {
        nxJson.tasksRunnerOptions.default.options.cacheableOperations = [
            ...nxJson.tasksRunnerOptions.default.options.cacheableOperations,
            'server',
        ];
        (0, devkit_1.updateNxJson)(tree, nxJson);
    }
}
exports.updateProjectConfig = updateProjectConfig;
