// @nrwl/angular/src/generators/utils/create-ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTsConfig = exports.extractTsConfigBase = void 0;
const devkit_1 = require("@nrwl/devkit");
const create_ts_config_1 = require("@nrwl/workspace/src/utils/create-ts-config");
const version_utils_1 = require("./version-utils");
var create_ts_config_2 = require("@nrwl/workspace/src/utils/create-ts-config");
Object.defineProperty(exports, "extractTsConfigBase", { enumerable: true, get: function () { return create_ts_config_2.extractTsConfigBase; } });
function createTsConfig(host, projectRoot, type, options, relativePathToRootTsConfig) {
    const majorAngularVersion = (0, version_utils_1.getInstalledAngularMajorVersion)(host);
    const json = {
        compilerOptions: {
            target: majorAngularVersion === 14 ? 'es2020' : 'es2022',
            useDefineForClassFields: false,
        },
        files: [],
        include: [],
        references: [
            {
                path: type === 'app' ? './tsconfig.app.json' : './tsconfig.lib.json',
            },
        ],
    };
    // inline tsconfig.base.json into the project
    if (options.rootProject) {
        json.compileOnSave = false;
        json.compilerOptions = Object.assign(Object.assign({}, create_ts_config_1.tsConfigBaseOptions), json.compilerOptions);
        json.exclude = ['node_modules', 'tmp'];
    }
    else {
        json.extends = relativePathToRootTsConfig;
    }
    (0, devkit_1.writeJson)(host, `${projectRoot}/tsconfig.json`, json);
}
exports.createTsConfig = createTsConfig;
